### PR TITLE
Add max-age=0 to the Cache-Control header to prevent stale content read from logged out users

### PR DESCRIPTION
**Pull request by bdauvergne**
As for cache-control we know that page protected by mod_mellon are
private and can vary depending on the current session. To prevent
navigator to show stale page after a logout we must set that the cached
page depends upon the value of the cookie header by setting the Vary
header. Without it navigators would be free to show page cached under a
connected session to anonymous users.

### CODE REVIEW

**Comment by olavmrk** [comment]
Is this really necessary? We set the `must-revalidate` flag, which basically tells the browser that it must contact the server before reusing the cached data. In addition we set the `private`-flag, which should prevent any shared caches from storing the data.

Have you seen a situation where the browser fetches the data from cache where that would stop if we provided the Vary-header?

**Comment by bdauvergne** [comment]
I have seen the problem. We have a simple WebDav server protected by mod_mellon à Entr'ouvert, we use it directly through the browser. without the line `Header append Vary: Cookie` added to the virtualhost configuration we always see the listing from the webdav after a logout and we are not automatically redirected to the idp (with `MellonEnable auth`). The private flag has no effect since the cached page comes from the browser. The must-revalidate does not work since we do not set a max-age or expire header and must-revalidate definition is:

        must-revalidate

        When the must-revalidate directive is present in a response received by a cache, that cache 
        MUST NOT use the entry after it becomes stale to respond to a subsequent request without first 
        revalidating it with the origin server

Without expire header or max-age cache-control flag  and with a Last-Modified header, the browser is free to choose a default expiration (at least, firefox does it https://blog.httpwatch.com/2008/10/15/two-important-differences-between-firefox-and-ie-caching/ ) and so must-revalidate will only be effectful when the cache expire.

So my problem has two solutions: setting max-age=0, I can only do it on my side, I cannot impose such policy to all mod_mellon users; but the fact is that whatever the cache expiration date is the cache is already stale because the page apparently depends upon an active session (if not why is it protected by mod_mellon); so the more logicial behaviour would be to indicate to the browser that the page (even cached) depends upon the Cookie value.

Without that you will have to explain to all your users in the README that they must set max-age=0 on the landing page following the logout of all their applications; it's possible that most php applications already use max-age=0 or no-cache, a reason why the problem is not seen more. But I think that an authenticated page should not be shown to an unauthenticated user whatever its expiration date.

**Comment by olavmrk** [comment]
Hi,

I see that I have been mistaken in the meaning of the `must-revalidate` flag -- I thought it always triggered a revalidation.

In this case, I think the correct behavior is to change the default Cache-Control header generated by mod_auth_mellon to read:

```
Cache-Control: private, must-revalidate, max-age=0
```

That encapsulates the default behavior we want; i.e. always check with the server before reusing a cached response.

I do not think adding a `Vary: Cookie` is the correct approach. It will not always work, e.g. when the logout is triggered through SOAP, or when the session expires on the SP.

**Comment by bdauvergne** [comment]
You are right that it will not work with SOAP, but SOAP logout was a bad idea. Do you want me to propose a patch with max-age=0 ? Reading different articles on stackoverflow it seems that must-revalidate, max-age=0 is equal to no-cache (but not on IE, IE never use the cache with max-age=0, firefox and chrome use it but always revalidate). So I would propose

  Cache-Controle: private, no-cache

**Comment by olavmrk** [comment]
> IE never use the cache with max-age=0

Actually, from reading a lot of different answers on Stack Overflow and elsewhere, my impression is the opposite: IE treats `no-cache` as `no-store`. Setting `max-age=0` on the other hand should allow it to cache the data but reuse it if allowed by the server.

(Unfortunately, it is hard to tell what information is up to date, and what information is based on ancient versions of Internet Explorer.)

I also think we should include the `must-revalidate` flag -- if the client cannot reach the server for some reason, they shouldn't use the old data.

So, to summarize, I think we want:

```
Cache-Control: private, must-revalidate, max-age=0
```

**Comment by olavmrk** [comment]
Hi,

I think you accidentally pushed the wrong branch to this pull request?

**Comment by bdauvergne** [comment]
Yep sorry, here is the branch for you.

**Comment by olavmrk** [comment]
Thanks! I have merged the patch now.
